cheatsheet.rst

```diff
@@ -16,3 +17,626 @@ as both a PDF and PNG.
+        options = {"shared": [True, False]}                # Defines available options and defaults. "shared" is a common
+                                                           # option which specifies whether a library is static or shared
+
+        default_options = {"shared": False}
+        requires = "RequiredLib/0.1@user/stable"           # Defines package requirements
+        build_requires = "tool_a/0.2@user/testing"         # Defines requirements that are only used when the package is
+                                                           # built. These should be build and test tools only.
+        generators = "cmake"                               # Generator for the package: specifies which build system
+                                                           # type will be generated
+
+        def source(self):                                                # Obtains the source code for the project
+            self.run("git clone https://github.com/conan-io/hello.git")  # self.run() executes any command in the native shell
+            ...
+
+        def build(self):                                                 # Responsible for invoking the build system
+            cmake = CMake(self)                                          # Helper classes are available for several build systems
+            ...
+            if self.options.myoption1:                                   # Specify a conditional build requirement
+                self.build_requires("zlib/1.2@user/testing")
+            self.run("bin/unittests")                                    # Run unit tests compiled earlier in the build() method
+
+        def package(self):                                               # Responsible for capturing build artifacts
+            self.copy("\*.h", dst="include", src="hello")                # self.copy() copies files from the cache to the project folder
+            ...
+
+        def package_info(self):                                          # Responsible for defining variables that are
+                                                                         # passed to package consumers, for example
+                                                                         # library or include directories
+            self.cpp_info.libs = ["hello"]                               # The cpp_info dictionary contains these variables
+            ...
+
+        def requirements(self):                                          # Responsible for specifying non-trivial requirements logic
+            if self.options.myoption2:                                   # Specify a conditional requirement
+                self.requires("RequiredLib2/0.3@user/stable")
+
+        def package_id(self):                                            # Responsible for changing the way the package
+                                                                         # ID is calculated from the default
```

Comment:
"while building this package only"? Because as far as I understand, you can affect the package ID modes of dependencies using this method, e.g. `self.info.requires[dep].recipe_revision_mode()`. Correct me if I'm wrong...
(Aside: what happens if `dep` defines a different package ID mode for itself, in its recipe? Does the mode defined in the dependent package win?)

Reply:
highly advanced question, don't know off hand, maybe not a good candidate for this "cheatsheet"

Reply:
Ok thanks. This sentence now reads "Responsible for overriding the way the package ID is calculated from the default, for this package only"